rir/src/compiler/analysis/global_redundancies.cpp

```diff
@@ -0,0 +1,122 @@
+#include "global_redundancies.h"
+
+namespace rir {
+namespace pir {
+  
+void GlobalRedundanciesAnalysis::apply(GRPartition& partition, Instruction* instruction) const {
+    // No operator
+    LdArg* loadArg = LdArg::Cast(instruction);
+    LdFun* loadFun = LdFun::Cast(instruction);
+    Instruction* loadVar = LdVar::Cast(instruction);
+    LdVarSuper* superLoadVar = LdVarSuper::Cast(instruction);
+    LdConst* loadConst = LdConst::Cast(instruction);
+
+    // Operator and 1 operand
+    Inc* inc = Inc::Cast(instruction);
+    IsObject* isObject = IsObject::Cast(instruction);
+
+    // Operator and 2 operands
+    Add* add = Add::Cast(instruction);
+    Gte* gte = Gte::Cast(instruction);
+    Lte* lte = Lte::Cast(instruction);
+    Mul* mul = Mul::Cast(instruction);
+    Div* division = Div::Cast(instruction);
+    IDiv* idiv = IDiv::Cast(instruction);
+    Mod* mod = Mod::Cast(instruction);
+    Colon* colon = Colon::Cast(instruction);
+    Pow* power = Pow::Cast(instruction);
+    Sub* sub = Sub::Cast(instruction);
+    Gt* gt = Gt::Cast(instruction);
+    Lt* lt = Lt::Cast(instruction);
+    Neq* neq = Neq::Cast(instruction);
+    Eq* eq = Eq::Cast(instruction);
+    LAnd* land = LAnd::Cast(instruction);
+    LOr* lor = LOr::Cast(instruction);
+
+    Phi* phi = Phi::Cast(instruction);
+    if (phi) {
+        /* It is a phi node. Create a phi eq-class if it is the first time
+         * Note that the original algorithm does not process phis but creates
+         * the phi equivalance classes at join points. For that it requires to
+         * be in CSSA. We workaround this by creating the phi equivalaence here.
+         * Check if we mantain semantics. 
+        */
+        if (!partition.count(instruction)) {
+            GREquivalenceClass klass;
+            klass.phis = new ValuePhiFunction(phi->arg(0).val(), phi->arg(1).val());
+            partition.insert({instruction, klass});
+        }
+        return;
+    }
+
+    partition.removeVariableFromClasses(instruction);
+
+    PIRAssignment assignment;
+    if (loadConst) {
+        assignment.setConstant(&loadConst->c);
+    } else if (loadArg) {
+        assignment.setIndex(loadArg->id);
+    } else {
+        ValueExpression::Expression expression;
+        if (loadFun || loadVar || superLoadVar) {
+            expression.operand1 = instruction->arg(0).first;
+        } else if (inc || isObject) {
+            expression.operand1 = instruction->arg(0).first;
+            expression.operation = &instruction->tag;
+        } else if (add || gte || lte || mul || division || idiv || mod || colon || 
+                power || sub || gt || lt || neq || eq || land || lor) {
+            expression.operand1 = instruction->arg(0).first;
+            expression.operand2 = instruction->arg(1).first;
+            expression.operation = &instruction->tag;            
+        }
+        assignment.setExpression(&expression);
```

Comment:
this won't work. You pass a pointer to the local variable `expression`. In C++ `Expression expression` creates a local variable on the stack. When the function returns, the stack frame vanishes and the local variable with it. Passing a pointer of a local variable and then keeping it alive longer than the function, is undefined. In this case it probably just results in a dangling pointer.

The way to go is probably to always pass Expressions by value. Since they should be really small structs (maybe 3 words big), I think it's ok.